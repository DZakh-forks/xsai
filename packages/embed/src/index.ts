import { clean, type CommonRequestOptions, requestUrl } from '@xsai/shared'

export type EmbedModel = 'all-minilm' | 'mxbai-embed-large' | 'nomic-embed-text' | ({} & string)

export interface EmbedOptions extends CommonRequestOptions<'embeddings'> {
  [key: string]: unknown
  input: string | string[]
  model: EmbedModel
}

export interface EmbedResponse {
  data: {
    embedding: number[]
    index: number
    object: 'embedding'
  }[]
  model: string
  object: 'list' | ({} & string)
  system_fingerprint?: string
  usage: EmbedResponseUsage
}

export interface EmbedResponseUsage {
  prompt_tokens: number
  total_tokens: number
}

export interface EmbedResult {
  embedding: number[]
  request: Request
  response: Response
  usage: EmbedResponseUsage
}

export const embed = async (options: EmbedOptions) => {
  const request = new Request(requestUrl(options.path ?? 'embeddings', options.base), {
    body: JSON.stringify(clean({
      ...options,
      abortSignal: undefined,
      base: undefined,
      headers: undefined,
      path: undefined,
    })),
    headers: {
      'Content-Type': 'application/json',
      ...options.headers,
    },
    method: 'POST',
    signal: options.abortSingal,
  })

  const response = await fetch(request)

  const json = await response.json() as EmbedResponse

  return {
    embedding: json.data[0].embedding,
    request,
    response,
    usage: json.usage,
  }
}

export default embed
