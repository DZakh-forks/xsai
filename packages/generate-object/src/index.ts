export { generateObject, generateObject as default, type GenerateObjectOptions, type GenerateObjectResult } from './utils/generate-object'
