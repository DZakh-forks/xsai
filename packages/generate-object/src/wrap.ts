import type { JsonSchema } from 'xsschema'

export const wrap = ({ $schema, ...schema }: JsonSchema): JsonSchema => ({
  properties: {
    elements: {
      items: schema,
      type: 'array',
    },
  },
  required: ['elements'],
  type: 'object',
})
