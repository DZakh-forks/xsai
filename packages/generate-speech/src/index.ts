import { type CommonRequestOptions, requestBody, requestHeaders } from '@xsai/shared'

export interface GenerateSpeechOptions extends CommonRequestOptions {
  [key: string]: unknown
  input: string
  /** @default `mp3` */
  responseFormat?: 'aac' | 'flac' | 'mp3' | 'opus' | 'pcm' | 'wav'
  /** @default `1.0` */
  speed?: number
  voice: string
}

/**
 * @experimental WIP
 * @param options generate speech options.
 * @returns audio array buffer
 */
export const generateSpeech = async (options: GenerateSpeechOptions): Promise<ArrayBuffer> =>
  await fetch(new URL('audio/speech', options.baseURL), {
    body: requestBody(options),
    headers: requestHeaders({
      'Content-Type': 'application/json',
      ...options.headers,
    }, options.apiKey),
    method: 'POST',
    signal: options.abortSignal,
  })
  // TODO: error handling
    .then(res => res.arrayBuffer())

export default generateSpeech
