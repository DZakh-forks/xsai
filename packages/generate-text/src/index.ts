import { clean, type CommonRequestOptions, objCamelToSnake, requestUrl } from '@xsai/shared'

import type { FinishReason, Message, TextGenerationModel, Tool, ToolChoice } from './types'

export interface GenerateTextOptions extends CommonRequestOptions<'chat/completions'> {
  [key: string]: unknown
  messages: Message[]
  model: TextGenerationModel
  toolChoice?: ToolChoice
  // tools?: Tool[]
}

export interface GenerateTextResponse {
  choices: {
    finish_reason: FinishReason
    index: number
    message: Message
  }[]
  created: number
  id: string
  model: string
  object: 'chat.completion'
  system_fingerprint: string
  usage: GenerateTextResponseUsage
}

export interface GenerateTextResponseUsage {
  completion_tokens: number
  prompt_tokens: number
  total_tokens: number
}

export interface GenerateTextResult {
  finishReason: FinishReason
  request: Request
  response: Response
  text: string
  usage: GenerateTextResponseUsage
}

export const generateText = async (options: GenerateTextOptions): Promise<GenerateTextResult> => {
  const request = new Request(requestUrl(options.path ?? 'chat/completions', options.base), {
    body: JSON.stringify(clean({
      ...objCamelToSnake(options),
      base: undefined,
      headers: undefined,
      path: undefined,
      stream: false,
      tools: (options.tools as Tool[] | undefined)?.map(tool => ({
        function: tool.function,
        type: 'function',
      })),
    })),
    headers: {
      'Content-Type': 'application/json',
      ...options.headers,
    },
    method: 'POST',
  })

  const response = await fetch(request)

  const json = await response.json() as GenerateTextResponse

  // TODO: FIXME: remove this
  console.log(JSON.stringify(json, null, 2))

  const { finish_reason, message } = json.choices[0]

  if (message.tool_calls) {
    const tool = (options.tools as Tool[]).find(tool => tool.function.name === message.tool_calls!.function.name)!
    const toolResult = await tool.execute(JSON.parse(message.tool_calls.function.arguments))
    const toolMessage = {
      content: toolResult,
      role: 'tool',
      tool_call_id: message.tool_calls.id,
    } satisfies Message

    return await generateText({
      ...options,
      messages: [
        ...options.messages,
        message,
        toolMessage,
      ],
    })
  }

  return {
    finishReason: finish_reason,
    request,
    response,
    text: message.content,
    usage: json.usage,
  }
}

export default generateText
