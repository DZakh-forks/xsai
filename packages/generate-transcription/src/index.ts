import { type CommonRequestOptions, requestHeaders } from '@xsai/shared'

export interface GenerateTranscriptionOptions extends CommonRequestOptions {
  // TODO: FIXME: file type
  file: unknown
  language?: string
  prompt?: string
  temperature?: string

  // response_format?: 'json' | 'text' | 'srt' | 'verbose_json' | 'vtt'
  // timestamp_granularities?: ('segment' | 'word')[]
}

export interface GenerateTranscriptionResult {
  text: string
}

/**
 * WIP, test failed
 */
export const generateTranscription = async (options: GenerateTranscriptionOptions): Promise<GenerateTranscriptionResult> => {
  const body = new FormData()

  body.append('model', options.model)
  // TODO: FIXME: file type
  body.append('file', options.file as any)

  return await (options.fetch ?? globalThis.fetch)(new URL('audio/transcriptions', options.baseURL), {
    body,
    headers: requestHeaders({
      'Content-Type': 'multipart/form-data',
      ...options.headers,
    }, options.apiKey),
    method: 'POST',
    signal: options.abortSignal,
  })
    .then(async (res) => {
      if (res.ok) {
        return res.json() as Promise<GenerateTranscriptionResult>
      }
      else {
        throw new Error(`Error(${res.status}): ${await res.text()}`)
      }
    })
    .then(({ text }) => ({ text: text.trim() }))
}

export default generateTranscription
