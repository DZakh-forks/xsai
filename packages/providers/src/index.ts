export * from './local'
export * from './remote'
export type * from './types'
export * from './utils'
