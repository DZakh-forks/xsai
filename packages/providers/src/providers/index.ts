export * from './google-generative-ai'
export * from './ollama'
export * from './openai'
