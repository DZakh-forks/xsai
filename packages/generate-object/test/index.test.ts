import { ollama } from '@xsai/providers'
import * as v from 'valibot'
import { describe, expect, it } from 'vitest'

import { generateObject } from '../src'

describe('@xsai/generate-object', () => {
  it('basic', async () => {
    const { object } = await generateObject({
      ...ollama.chat('llama3.2'),
      messages: [
        {
          content: 'You are a helpful assistant.',
          role: 'system',
        },
        {
          content: 'Generate a sous-vide steak recipe.',
          role: 'user',
        },
      ],
      schema: v.object({
        recipe: v.object({
          ingredients: v.array(v.object({ amount: v.string(), name: v.string() })),
          name: v.string(),
          steps: v.array(v.string()),
        }),
      }),
      seed: 39,
    })

    expect(object.recipe).toMatchSnapshot()
  }, 60000)
})
