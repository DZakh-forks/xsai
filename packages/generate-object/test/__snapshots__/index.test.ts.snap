// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@xsai/generate-object > basic 1`] = `
{
  "ingredients": [
    {
      "amount": "2",
      "name": "steak",
    },
    {
      "amount": "1/4 cup",
      "name": "olive oil",
    },
    {
      "amount": "",
      "name": "garlic powder",
    },
    {
      "amount": "",
      "name": "paprika",
    },
    {
      "amount": "1 tablespoon",
      "name": "kosher salt",
    },
    {
      "amount": "1 teaspoon",
      "name": "black pepper",
    },
  ],
  "name": "sous vide steak recipe",
  "steps": [
    "Step 1: Prepare your steak. Bring the steak to room temperature before cooking, about 30 minutes.",
    "Step 2: Seasoning Mix a dry blend of garlic powder and paprika. Add salt and pepper according to taste to season. Rub this mix evenly across steak.",
    "Step 3: Heat Sous, Seal with plastic wrap and Place Steak at desired Temperature (Between 120°F - 160 °F), Let it cook for 1 hour before searing with hot pan. You may remove the wrapping to allow juices from coming out during sear",
    "Step 4: Pan-searing Add steak without wrapping, sauté in hot pan on high heat (with a small amount of olive oil) until desired crust.",
    "Step 5: Final touches Remove Pan-cooked steak and immediately whisk 2 tablespoons butter and one tablespoon juice.",
  ],
}
`;
